src/participant.rs

```diff
@@ -25,16 +25,87 @@ struct ProgressIndex {
     sid: Identifier,
 }
 
+/// Possible outcomes from processing one or more messages.
+pub(crate) enum ProcessOutcome<O> {
+    // Didn't finish processing the message; we need more inputs to continue
+    Unprocessed,
+    // The message was processed successfully but the subprotocol isn't done
+    Processed(Vec<Message>),
+    // The subprotocol is done for this participant but there are still messages to send to others
+    TerminatedForThisParticipant(O, Vec<Message>),
+    // The entire subprotocol is done and there are no more messages to send
+    FullyTerminated(O),
+}
+
+impl<O> ProcessOutcome<O> {
+    /// Create a `ProcessOutcome` from an optional output and a set of outgoing
+    /// messages.
+    pub(crate) fn from(output: Option<O>, messages: Vec<Message>) -> Self {
+        match (output, messages.len()) {
+            (None, 0) => Self::Unprocessed,
+            (None, _) => Self::Processed(messages),
+            (Some(o), 0) => Self::FullyTerminated(o),
+            (Some(o), _) => Self::TerminatedForThisParticipant(o, messages),
+        }
+    }
+
+    /// Extract the outoging messages from the `ProcessOutcome`.
+    ///
+    /// This method drops the outcome, if it exists, so it's no longer
+    /// accessible.
+    pub(crate) fn into_messages(self) -> Vec<Message> {
+        match self {
+            Self::Unprocessed | Self::FullyTerminated(_) => Vec::new(),
+            Self::Processed(messages) => messages,
+            Self::TerminatedForThisParticipant(_, messages) => messages,
+        }
+    }
+
+    /// Convert the `ProcessOutcome` into its consituent parts.
```

Comment:
Typo.

Reply:
I actually sounded out this word because it looked weird to me, but I didn't spot the typo :upside_down_face: